Remove `indexing_complete` when removing policy

gwbrown opened the pull request:
Leaving `index.lifecycle.indexing_complete` in place when removing the
lifecycle policy from an index can cause confusion, as if a new policy
is associated with the policy, rollover will be silently skipped.
Removing that setting when removing the policy from an index makes
associating a new policy with the index more involved, but allows ILM to
fail loudly, rather than silently skipping operations which the user may
assume are being performed.

---

Context: After synchronous discussion earlier today, @talevy, @eskibars, and I decided that failing loudly when we might possibly be able to succeed is better than being silently surprising, especially when the fix is relatively straightforward.

elasticmachine commented:
Pinging @elastic/es-core-features

gwbrown commented:
@martijnvg I requested your review here partially to double-check my thinking here: This setting will be used (in combination with other signals) to indicate to ILM that it is safe to de-follow a following index.  I believe the only case when removing this setting when removing a policy would have an impact is if the policy is removed from an index after it rolls over (and so is actually finished indexing), but a follower has not yet caught up.  In this case, the following index will *not* have this setting set, and so be stuck waiting for this setting to be set.  The fix is simply to set this setting manually.  Correct?

colings86 commented on x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/indexlifecycle/IndexLifecycleRunnerTests.java:
```diff
@@ -1134,6 +1134,32 @@ public void testRemovePolicyForIndexIndexInUnsafe() {
+
+        assertTrue(failedIndexes.isEmpty());
+        assertIndexNotManagedByILM(newClusterState, index);
+    }
```
Could we add a test that gets an index past rollover, runs remove policy, then sets up ILM on the index again, waits for the rollover step to error, sets indexing complete and runs the retry API? I know this sounds convoluted but I want to make sure we have a test for the scenario when a user want to re-add ILM (because maybe they made a mistake by removing it) and ensure things work as expected.

gwbrown commented:
@colings86 I've added the test you asked for. In the process of writing the test, I ran across a bug: Instead of moving to the error step at https://github.com/elastic/elasticsearch/pull/36620/files#diff-5de660b10fd8979ff41a9ff8d850aaecR569, execution of the policy for this index would start checking for rollover, because we didn't check that the index was actually the write index for the alias.

I've added that check in `WaitForRolloverReadyStep`, and adjusted the order of the checks in that step, which is necessary to accommodate both new-style aliases with write indices set and legacy-style aliases which only point to one index at a time.

Because there's now a slight difference in how those are handled, I've tweaked `createIndexWithSettings` to randomize between new-style and legacy-style aliases - I've run all tests both ways and they pass.

It might be worth pulling those changes out into a separate PR, but this PR won't pass without those changes.

colings86 commented on x-pack/plugin/ilm/qa/multi-node/src/test/java/org/elasticsearch/xpack/indexlifecycle/TimeSeriesLifecycleActionsIT.java:
```diff
@@ -531,6 +531,58 @@ public void testDeletePolicyInUse() throws IOException {
+            "  \"index.lifecycle.indexing_complete\": true\n" +
+            "}");
+        client().performRequest(setIndexingCompleteRequest);
+
```
Should we check to make sure the index reaches an error step before running retry?
